### PR TITLE
[firtool] btor2 integration

### PR DESCRIPTION
**Pull request by dobios**
Hi,

This PR adds a `btor2` target to `firtool`.
The goal is to expose this directly in Chisel later.

The one thing I might add here is that it might also be useful to add the `arc::InlineModulesPass` to the pipeline, as module instances need to be inlined in order to get something that will work in `btor2`, but I wanted to get some opinions first before integrating some pass from an unrelated dialect into the pipeline.

Let me know what you think!

### CODE REVIEW

File: tools/firtool/firtool.cpp
Comment (uenoku) [inline tools/firtool/firtool.cpp:418]: Can we hoist LowerFIRRTLToHW and hwPassPlugin part?

**Review by uenoku — COMMENTED** [review]
Can we add a simple test for this pipeline? It would be also great if there is an integration test to check/run emitted btor2 with btormc or something.

**Comment by dtzSiFive** [comment]
Woohoo!!

Please register the pass with the other conversion passes so MLIR debugging bits work with this, I'm on mobile but can be more useful about what I mean if unclear!

**Comment by dobios** [comment]
> Woohoo!!
> 
> Please register the pass with the other conversion passes so MLIR debugging bits work with this, I'm on mobile but can be more useful about what I mean if unclear!

I'm not quite sure that I understand what you mean by this, is this something specific to firtool or more general?

File: tools/firtool/firtool.cpp
Comment (dobios) [reply tools/firtool/firtool.cpp:418]: yeah ofc

**Comment by dtzSiFive** [comment]
@dobios https://github.com/llvm/circt/blob/f424ea325c4cbd68e5564b6d2ba9cab0818c77d1/tools/firtool/firtool.cpp#L718 , need passes registered when parsing command line options for some bits to work.

**Comment by dobios** [comment]
> @dobios
> 
> https://github.com/llvm/circt/blob/f424ea325c4cbd68e5564b6d2ba9cab0818c77d1/tools/firtool/firtool.cpp#L718
> , need passes registered when parsing command line options for some bits to work.

@dtzSiFive Got it, registered the pass with the other conversion passes!

**Comment by maerhart** [comment]
If people agree, we could also move the `InlineModulesPass` to HW transforms, as it only operates on HW operations and doesn't need a dependency on Arc.

**Comment by fabianschuiki** [comment]
> If people agree, we could also move the `InlineModulesPass` to HW transforms, as it only operates on HW operations and doesn't need a dependency on Arc.

That sounds like a good idea! Maybe `FlattenModulesPass` to indicate that it doesn't make any decisions about when to inline, it just inlines everything.

**Review by fabianschuiki — APPROVED** [review]
This looks good to me! It would be great to have a tiny test that sanity checks whether you get _something_ when you ask for this btor output. I'm not sure if we do this for some of the other formats, but it would probably be good to do so. There should be a firtool test directory where you could add something like `target-btor.mlir` to check this new emission option.

**Comment by uenoku** [comment]
> That sounds like a good idea! Maybe FlattenModulesPass to indicate that it doesn't make any decisions about when to inline, it just inlines everything.

Can we instead add a callback to decide whether a module should be inlined? This approach would be useful when we want to inline only small modules etc.

**Comment by dtzSiFive** [comment]
> Can we instead add a callback to decide whether a module should be inlined? This approach would be useful when we want to inline only small modules etc.

MLIR has an inliner interface -- maybe something like that, vs pass having a callback?

**Comment by fabianschuiki** [comment]
We could definitely try to reuse the inliner interface, or model something after that. Do we have a use case for that at the moment? Might make sense to simply do the move to unblock some work, and then add the interface and per-module inlining control ontop of that. It would be nice to have a proper inliner 😃.

**Comment by uenoku** [comment]
> Ah, not question of legality but of intent (hence as an option/parameter/callback to the pass).

Yes probably the inliner cost model is not uniform so I think it's nicer to set a cost model aside from dialect interface. Checked MLIR inliner and the current design is to separate inlining logic into a helper struct which takes a predicate (https://discourse.llvm.org/t/inliner-cost-model/2992/13 and https://github.com/llvm/llvm-project/pull/84258) so probably we can do the similar thing. 

Anyway I think the PR looks good to me if there is a simple test for it :+1:

**Comment by dtzSiFive** [comment]
Sounds great! Thanks for the links and explanation.  Simply moving first seems like the way to go, agreed :+1:.

**Comment by dobios** [comment]
I'm waiting to get #6964 merged in before merging this one, as it would break if modules aren't inlined beforehand.

**Comment by sequencer** [comment]
cc @SpriteOvO this should be included in the PanamaBinding in chisel.

**Pull request merged**
